llama-index-integrations/llms/llama-index-llms-openvino-genai/llama_index/llms/openvino_genai/base.py

```diff
@@ -0,0 +1,408 @@
+import logging
+from typing import Any, Callable, Optional, Sequence, Union
+
+from llama_index.core.base.llms.types import (
+    ChatMessage,
+    ChatResponse,
+    ChatResponseGen,
+    CompletionResponse,
+    CompletionResponseGen,
+    LLMMetadata,
+)
+from llama_index.core.bridge.pydantic import Field, PrivateAttr
+from llama_index.core.callbacks import CallbackManager
+from llama_index.core.llms.callbacks import (
+    llm_chat_callback,
+    llm_completion_callback,
+)
+from llama_index.core.llms.custom import CustomLLM
+from llama_index.core.base.llms.generic_utils import (
+    completion_response_to_chat_response,
+    stream_completion_response_to_chat_response,
+    messages_to_prompt as generic_messages_to_prompt,
+)
+from llama_index.core.prompts.base import PromptTemplate
+import queue
+import openvino_genai
+import openvino as ov
+from threading import Event, Thread
+
+logger = logging.getLogger(__name__)
+
+
+class OpenVINOGenAILLM(CustomLLM):
+    r"""OpenVINO GenAI LLM.
+
+    Examples:
+        `pip install llama-index-llms-openvino-genai`
+
+        ```python
+        from llama_index.llms.openvino_genai import OpenVINOgenAILLM
+
+        llm = OpenVINOGenAILLM(
+            model_path=model_path,
+            device="CPU",
+        )
+
+        response = llm.complete("What is the meaning of life?")
+        print(str(response))
+        ```
+    """
+
+    model_path: str = Field(
+        default=None,
+        description=("The model path to use from local. "),
+    )
+    system_prompt: str = Field(
+        default="",
+        description=(
+            "The system prompt, containing any extra instructions or context. "
+            "The model card on HuggingFace should specify if this is needed."
+        ),
+    )
+    query_wrapper_prompt: PromptTemplate = Field(
+        default=PromptTemplate("{query_str}"),
+        description=(
+            "The query wrapper prompt, containing the query placeholder. "
+            "The model card on HuggingFace should specify if this is needed. "
+            "Should contain a `{query_str}` placeholder."
+        ),
+    )
+    device: str = Field(
+        default="auto", description="The device to use. Defaults to 'auto'."
+    )
+    is_chat_model: bool = Field(
+        default=False,
+        description=(
+            LLMMetadata.model_fields["is_chat_model"].description
+            + " Be sure to verify that you either pass an appropriate tokenizer "
+            "that can convert prompts to properly formatted chat messages or a "
+            "`messages_to_prompt` that does so."
+        ),
+    )
+
+    config: str = Field(
+        default=None,
+        description=("The LLM generation configurations."),
+    )
+
```

Comment:
Generally, everything looks good, thanks.

I have only questions about some advanced usage:

- if some additional parameters needed for pipeline initialization (ov_config, scheduler_config, draft_model for speculative decoding, adapter_config for LoRA) how can we provide them?
- what if the model is provided in the original pytorch format, should we call optimum-intel to convert it on the fly or it should be beyond integration?

Reply:
Hi I added `config`  and `**kwargs` to receive these parameters.

For second question, I think user should convert their model through optimum-cli offline, and I also add guideline for this part in documents.